Add support for CMake build system

wrobelda opened the pull request:
Any suggestions are most welcome.

A separate Pull Request for the inclusion of FindOpenSP module with CMake was filed with Kitware:
https://gitlab.kitware.com/cmake/cmake/-/merge_requests/7470

Todo:
- [x] `HAVE_GCC_VISIBILITY_EXTS` vs `CMAKE_CXX_VISIBILITY_PRESET`
- [x] `getopt()`/`getopt_long()`  situation on Windows (https://github.com/libofx/libofx/pull/50) and other platforms:
   - we can just do nothing, and assume Windows users would use vcpkg, which has a drop-in `getopt-win32` replacement available if needed. The code would not be compile with vanilla MSVC out of the box, though.
   - we can replace the `getopt()`/`getopt_long()` usage with the BSD-licensed https://github.com/kubo/ya_getopt 
   - we can drop `getopt_long()` and rely on the included copy of `getopt()`. After all, we only use it for the tools which are mere additions to the library itself. 
- [x] should the support for regenerating `cmdline.c`/`cmdline.h` be dropped and should we rely on a checked-in copy of it? Is there any advantage of regenerating them locally anymore?
- [x] static/shared build tests
   - Windows shared build fails with `[build] LINK : fatal error LNK1104: cannot open file '..\lib\Debug\libofx.lib' `
- [x] vcpkg’s iconv seems to actually be Unix compatible, regarding the `“#ifdef _WIN32 //Win32's iconv implementation takes a const char** inchar”` line, which means that a mere `#ifdef _WIN32` ~~might not be~~ is not enough to choose the right code path
   - this isn't specific to any platform, but rather to the iconv standard, see
     - https://stackoverflow.com/questions/45938990/why-does-the-iconv-function-need-a-non-const-inbuffer
     - https://github.com/win-iconv/win-iconv/issues/8  
     - https://github.com/win-iconv/win-iconv/issues/25
     - https://github.com/win-iconv/win-iconv/blob/master/iconv.h
     - https://github.com/Microsoft/vcpkg/issues/4904
     - https://sourceware.org/bugzilla/show_bug.cgi?id=2962
   - need to patch that code and ideally detect const in buffer or possibly add a configurable switch
- [x] consider turning `getopt/CMakeLists` into a CMake Find module
- [x] ~~`install(TARGETS tool name RUNTIME DESTINATION bin)` does not copy the `osp.dll` on Windows (the lib gets copied, though). Not sure if this has something to do with the `install()` syntax itself or my `FindOpenSP` module that may need some extra magic to detect how the build was build — even though it does `add_library(OpenSP::OpenSP UNKNOWN IMPORTED)`, and the `UNKNOWN` is how CMake manual [advertises](https://cmake.org/cmake/help/latest/command/add_library.html#imported-libraries) its use for find modules~~  **EDIT:** works as intended, would require an additional `install(RUNTIME_DEPENDENCY_SET)` to also copy dependent's DLLs
- [x] output on Windows is garbled up (also #62)
- [ ] go through automake-generated `config.h` to see if anything of importance was missed when porting over to `config.h.cmake` (@cstim)
- [x] the `getopt` implementation is faulty, doesn't handle more than 1 parameter at a command line. Need to find a yet another replacement, I guess.
- [x] get getopt code to properly detect vcpkg-provided getopt
- [x] add support for building out of the box with vcpkg
   - ofxconnect won't build right now, due to `vcpkg` having `libxmlpp` at current 5.x version, while ofxconnect expects 2.6 ABI. 
- [x] update README
- [ ] get tests/checks running. Is ofxconnect's `test-privateserver.sh` still functional? 

Fixes  #46

TODO post-merge, during removal of autotools:
* go through `CMakeLists.txt`, see the TODOs there
* rename `README` to `README.md`
* remove autotools-specific instructions from README
* potentially remove Direct Connect testing code and its mentions in README

wrobelda committed:
Initial CMakeLists.txt and FindOpenSP.cmake

These were initially provided by Cristian Oneț
<onet.cristian@gmail.com> for KMyMoney build
system:
https://github.com/KDE/kmymoney/commits/master/cmake/modules/FindLibOfx.cmake

wrobelda committed:
Modernize FindOpenSP.cmake

wrobelda committed:
gengetopt ofxdump's cmdline and keep it

wrobelda committed:
Use version variables that match automake's

wrobelda committed:
Fix OpenSP includes

wrobelda committed:
Add config.h.cmake stub

wrobelda committed:
Properly set _WIN32 on Windows

wrobelda committed:
Move ofxdump CMake code to its own dir

wrobelda committed:
gengetopt ofxconnet's cmdline and keep it

wrobelda committed:
Add feature summary

wrobelda committed:
Make ofxdump a feature

wrobelda committed:
Add support for building ofxconnect

wrobelda committed:
Add support for building ofx2qif

wrobelda committed:
Set up CMAKE_BUILD_TYPE

wrobelda committed:
Parse libofx.h.in

wrobelda committed:
Search for OpenSP include insensitively; reformat

wrobelda committed:
Update .gitignore

wrobelda committed:
Simplify /dtd directory installation

jralls commented on CMakeLists.txt:
```diff
@@ -0,0 +1,67 @@
+
+install(TARGETS ofx ofxdump RUNTIME DESTINATION bin ARCHIVE DESTINATION lib LIBRARY DESTINATION lib)
+install(FILES inc/libofx.h DESTINATION include/libofx)
+install(FILES dtd/opensp.dcl DESTINATION "${CMAKE_PREFIX_PATH}/share/libofx/dtd")
```
You inadvertently fix this later, but for your edification: CMAKE_PREFIX_PATH is a variable that Cmake uses to find dependencies. CMAKE_INSTALL_PREFIX tells Cmake where to install build products.
Edit: Note as well that CMAKE_PREFIX_PATH is allowed to be a list of paths.

wrobelda replied:
This was copied from the original CMakeList I am basing this on (https://invent.kde.org/packaging/craft-blueprints-kde/-/blob/2a47ef6c3ef1b1ad1aa3695080ea89cd86c4a6ee/libs/libofx/libofx-0.9.13-20180505-1.diff), and was, in fact, fixed advertently. 

I appreciate the feedback, though.

jralls commented on CMakeLists.txt:
```diff
@@ -0,0 +1,67 @@
+# SPDX-FileCopyrightText: 2009 Cristian Oneț <onet.cristian@gmail.com>
```
libofx doesn't use SPDX, nor does an SPDX line comply with the requirements of copyright law or the terms of the Gnu General Public License.  See https://www.gnu.org/licenses/gpl-howto.html for instructions on how to correctly document the copyright and GPL licensing in files.

Prior to [your changes](https://github.com/KDE/kmymoney/commit/09172a150bd2185889b9887a4406a6374c0726a0) the copyright and license of this file was
```
# Copyright (c) 2006, Alexander Neundorf, <neundorf@kde.org>
# Copied from FindLibXslt.cmake, 2009, Guillaume DE BURE, <gdebure@yahoo.com>
#
# Redistribution and use is allowed according to the terms of the BSD license.
# For details see the accompanying COPYING-CMAKE-SCRIPTS file.
```
Aside from screwing up the copyright claim in that change you discarded the license statement, violating Alexander Neudorf's terms. Contrary to your claim in the commit message, this file was not created by Christian Onet, he just lifted it from somewhere else and committed it to KMyMoney.

Edit: 
Sorry, I put this comment in the wrong place, the copyright and BSD license comments apply to FindOpenSP.cmake.

It's not common to assert copyright on CMakeLists.txt, but that aside it doesn't seem to me that assigning copyright here to Mr. Connet is appropriate. Even if you copied heavily from the KMyMoney CMakeLists.txt that's hardly the sole work of Mr. Connet and that file has no copyright assertion.

wrobelda replied:
I spent a good chunk of time to dig down to the first author of the file, which I actually first noticed used as part of a patch here: https://invent.kde.org/packaging/craft-blueprints-kde/-/blob/master/libs/libofx/libofx.py

I eventually stumbled upon the same code in KMyMoney, but actually failed to notice the header mentions the name author different than that of the committer. Do note that time I was working with the `craft-blueprints-kde`'s copy of the file from the beginning, which already had had the license header removed, so I simply assumed Christian was the author himself.

So, while I appreciate pointing this out, jumping into conclusion and implying I intentionally discarded the license and violated author's rights is a bit much. I mean, we all (at least I do) it for free, so what potential gain would I have from changing the ownership of a 15 y/o file from one 3rd party to another 3rd party?

wrobelda replied:
Also,

> Contrary to your claim in the commit message, this file was not created by Christian Onet, he just lifted it from somewhere else and committed it to KMyMoney.

Christian has also subsequently amended the file himself, so absolutely has right to be mentioned there. 

Lastly, I rewrote `FindOpenSP` module from scratch, so the license no longer applies there. And the `libofx` CMakeList.txt is pretty much written from scratch by that point, too, since it originally was a hack to compile code under Windows, but I updated its copyrights nonetheless.

jralls replied:
> So, while I appreciate pointing this out, jumping into conclusion and implying I intentionally discarded the license and violated author's rights is a bit much.

Sorry, I didn't mean to imply that you'd done it intentionally.  But...
> which already had had the license header removed

isn't right. In fact *it's still there* in the [KMyMoney version](https://github.com/KDE/kmymoney/blob/aa17abb7e9e09b4f811522fcb7135e91a0f4189b/cmake/modules/FindLibOfx.cmake#L9). You lost it when you copied the file to here.

wrobelda replied:
> Sorry, I didn't mean to imply that you'd done it intentionally. But...
> 
> > which already had had the license header removed
> 
> isn't right. In fact _it's still there_ in the [KMyMoney version](https://github.com/KDE/kmymoney/blob/aa17abb7e9e09b4f811522fcb7135e91a0f4189b/cmake/modules/FindLibOfx.cmake#L9). You lost it when you copied the file to here.

As I explained, I did _not_ use that copy as my starting point. I used this: https://invent.kde.org/packaging/craft-blueprints-kde/-/blob/2a47ef6c3ef1b1ad1aa3695080ea89cd86c4a6ee/libs/libofx/libofx-0.9.13-20180505-1.diff

wrobelda committed:
WIP: getopt win32

wrobelda committed:
Move tool handling code to their respective CMakeLists.txt

jralls commented on CMakeLists.txt:
```diff
@@ -0,0 +1,67 @@
+)
+
+add_definitions(
+    -D__WIN32__
```
__WIN32__ is set by the compiler. It shouldn't be set in CMakelLists.txt. Worse, setting it here prevents Cmake being used in on non-Win32 OSes.

wrobelda replied:
Indeed, this is a leftover from the original file, but I haven't tested it under MSVC yet, so I left as is.

jralls replied:
Actually you didn't, you changed it to `-D_WIN32=1` and that is just as wrong, it's just the MSVC version of wrong instead of the GCC one. Those macros are for the compiler to tell the preprocessor what compiler is compiling for what target. You shouldn't interfere by setting them in the build system.

wrobelda committed:
Move libofx code to its own CMakeLists.txt

wrobelda committed:
Copy all tools to /bin

jralls commented on CMakeLists.txt:
```diff
@@ -0,0 +1,67 @@
+
+add_definitions(
+    -D__WIN32__
+    -DMAKEFILE_DTD_PATH="${CMAKE_PREFIX_PATH}/share/libofx/dtd/"
```
`CMAKE_INSTALL_PREFIX`, not `CMAKE_PREFIX_PATH`. The latter may be a ';' delimited list of paths that would seriously blow up the `DTD_SEARCH_PATH`.

[`target_compile_definitions`](https://cmake.org/cmake/help/latest/command/target_compile_definitions.html) is the preferred modern-cmake way of setting preprocessor symbols to `add_definitions`.

wrobelda replied:
Indeed, this is a leftover from the original file, but I haven't tested it under MSVC yet, so I left as is.

jralls commented on CMakeLists.txt:
```diff
@@ -0,0 +1,67 @@
+# We would like it to be shared, but libofx does not export symbols manually
+# thus it causes troubles on MSVC
+#add_library(ofx SHARED ${SRC_SOURCES})
+add_library(ofx ${SRC_SOURCES})
```
Adding `dllexport` decorators would be a better solution, but absent that this should be
```
if(MSVC)
   add_library(ofx ${SRC_SOURCES})
else()
   add_library(ofx SHARED $SRC_SOURCES})
endif()

wrobelda replied:
Disagree, this would enforce dynamic linkage without possibility to override it. The library works fine compiled statically, which is useful for `vcpkg` which supports static linkage on macOS and Linux for the time being. 

From the docs:


> STATIC, SHARED, or MODULE _may be_ given to specify the type of library to be created. (...) If no type is given explicitly the type is STATIC or SHARED based on whether the current value of the variable [BUILD_SHARED_LIBS](https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html#variable:BUILD_SHARED_LIBS) is ON.

jralls replied:
I didn't write the comment about wanting it to be shared but that not working on MSVC, you did. If that's not the case, remove the comment.

wrobelda replied:
> you did.

Again, this is a leftover, and you yet again jump into such conclusion. All of this is a WIP, marked as Draft, so I would have hope for it to be clear by now that assuming I wrote everything here is counterproductive.

wrobelda replied:
I just realized I fixed that already some **20 commits ago, 6 DAYS AGO** — even before I filed this PR. So you're seriously going through the history of my commits of this WORK IN PROGRESS DRAFT and commenting on the things that *were there* but no longer are? 

What's the point of that, John? Cause this is anything but in the spirit of Open Source contribution, and a waste of our time.

jralls replied:
I usually do code reviews commit-by-commit for non-trivial PRs, it works better with Github than reviewing on the Files changed tab. If you don't want someone looking over your shoulder why show your work? You can `git rebase` away all of the messy stuff before making it public. You might also note that comments on code that you've changed has a gold Outdated label right at the top.

I'm not sure what any of that has to do with the spirit of Open Source contribution, though I'll stipulate for you that it may end up taking more time and that you might consider that wasted time.

jralls commented on CMakeLists.txt:
```diff
@@ -0,0 +1,67 @@
+)
+
+file(GLOB COEFF_HDRS inc/libofx.h lib/*.hh )
+file(GLOB SRC_SOURCES lib/*.cpp ${COEFF_HDRS})
```
Very bad practice. List the files, and use more expressive variable names like `ofx_HEADERS` and `ofx_SOURCES`.

wrobelda replied:
Also a leftover from the original file. Haven't "prettified" the variables yet, editing as I go.

EDIT: actually, GLOB is mine, but this was for convenience. Will rewrite once everything is ready.

jralls commented on CMakeLists.txt:
```diff
@@ -0,0 +1,67 @@
+    set(ofxdump_SRCS ${ofxdump_SRCS} lib/messages.cpp lib/getopt.c lib/getopt1.c)
+endif(MSVC)
+
+add_definitions(-DCMDLINE_PARSER_PACKAGE=\"ofxdump\" -DCMDLINE_PARSER_PACKAGE_NAME=\"ofxdump\" -DCMDLINE_PARSER_VERSION=\"${VERSION}\")
```
Using `add_definitions` applies globally to all targets. `target_compile_definitions` will cause these to be defined only for ofxdump.

wrobelda replied:
Also a leftover from the original file. I haven't gotten to actually understand the purpose behind these, so left as is for now.

jralls replied:
`add_definitions` just pastes the argument string into the compiler invocation, which is why you need to escape the quotes so that cmake doesn't eat them. `target_compile_definitions` does a little preprocessing for you so you don't need to escape the quotes or provide the `-D`; it also applies the arguments only to the specified target instead of globally to all targets the way `add_definitions` does.

IOW it's telling the build system to say `gcc -c -DCMDLINE_PARSER_PACKAGE="ofxdump" -o target.o -c source.c`.

wrobelda committed:
Fix config.h.cmake substitution

Use find_package() to find Iconv module

jralls commented on ofxconnect/.gitignore:
```diff
@@ -6,6 +6,4 @@
 Makefile.in
 ofxconnect
-cmdline.c
-cmdline.h
```
Something's squirrelly: This duplicates f3f8837e5f1 but doesn't cause any conflicts.

wrobelda replied:
This is for `ofxconnect`, the other one is for `ofxdump`

jralls replied:
Ah, got it.

jralls commented on lib/ofx_preproc.cpp:
```diff
@@ -70,6 +70,7 @@ const char *DTD_SEARCH_PATH[DTD_SEARCH_PATH_NUM] =
 #endif
   "/usr/local/share/libofx/dtd",
   "/usr/share/libofx/dtd",
+  "../dtd",
```
".." is relative to $PWD, not the lib source directory. Even if someone is being really stupid and building in the source dir they're probably going to run from the libofx root rather than one of the subdirs.

wrobelda replied:
This was temp to speed up hacking, I dropped it.

jralls commented on lib/getopt/getopt.c:
```diff
@@ -0,0 +1,973 @@
+Version: 1.0
+Comment: Supports getopt, getopt_long, and getopt_long_only
+and POSIXLY_CORRECT environment flag
+License: LGPL
```
Requires adding a copy of the LGPL of the correct version in the root directory. You may need to contact the author to find out what version that is, but if there's a copy in the place you lifted this from you can reasonably assume that's the right version.

OTOH vcpkg includes a getopt-win32 package. Why not use that?

wrobelda replied:
> OTOH vcpkg includes a getopt-win32 package. Why not use that?

That's exactly what [6228bd8](https://github.com/libofx/libofx/pull/63/commits/6228bd8ce3ee9e65cf63be1b18a843eea7a6eadc) does, but it's a WIP.

jralls replied:
I mean make the getopt-win32 package a dependency when building with MSVC instead of building it in here.

wrobelda replied:
> I mean make the getopt-win32 package a dependency when building with MSVC

Not sure how you think this would work, but `vcpkg` is not part of `MSVC`. Relying on `vcpkg` would mean compilation with pure MSVC wouldn't be possible. Refer to my first comment in this PR for context.

jralls replied:
Your original comment raises the same points but doesn't elaborate on the why's .  I'm trying to elicit the reasoning. Let's start with is compiling a stand-alone libofx with MSVC a worthwhile goal and if so, why? Do you know of some GPLV2 compliant project that compiles with only MSVC and might benefit from libofx?

If that's the case there's a fourth alternative to your three: Write a cmake FindGetoptWin32.cmake and let the user who wants to build on MSVC get the code and build it separately.

jralls replied:
> You may need to contact the author to find out what version that is

The [vcpkg item](https://vcpkg.info/port/getopt-win32) points to a [GitHub repository](https://github.com/libimobiledevice-win32/getopt) that points to [the original article](https://www.codeproject.com/Articles/157001/Full-getopt-Port-for-Unicode-and-Multibyte-Microso) that has at the top a link to [the license](http://www.opensource.org/licenses/lgpl-3.0.html). It's LGPLV3. That might be a deal breaker for @benoitg because including that source here [makes libofx GPLV3](https://www.gnu.org/licenses/gpl-faq.html#AllCompatibility) instead of the current GPLV2+.

wrobelda replied:
> > You may need to contact the author to find out what version that is
> 
> The [vcpkg item](https://vcpkg.info/port/getopt-win32) points to a [GitHub repository](https://github.com/libimobiledevice-win32/getopt) that points to [the original article](https://www.codeproject.com/Articles/157001/Full-getopt-Port-for-Unicode-and-Multibyte-Microso) that has at the top a link to [the license](http://www.opensource.org/licenses/lgpl-3.0.html). It's LGPLV3. That might be a deal breaker for @benoitg because including that source here [makes libofx GPLV3](https://www.gnu.org/licenses/gpl-faq.html#AllCompatibility) instead of the current GPLV2+.

Indeed. The alternative is MIT-licensed https://github.com/takamin/win-c

wrobelda replied:
> Let's start with is compiling a stand-alone libofx with MSVC a worthwhile goal and if so, why?

Because of the tools provided by the libofx itself (ofxdump, ofxconnect, ofx2qif). The former two are, after all, the __only__ components that depend on `getopt`. And since they also intend to test the library, it makes sense to me that we provide a "one stop shop" solution to build them. If the library alone was to be considered, we wouldn't be discussing this at all.

> if that's the case there's a fourth alternative to your three: Write a cmake FindGetoptWin32.cmake and let the user who wants to build on MSVC get the code and build it separately.

Indeed, that's basically the "we can just do nothing" option. Whether a user decides to build their own copy or resorts to vcpkg, it would be up to them. In such case, I would convert the `getopt/CMakeLists.txt` to `FindGetoptWin32.cmake`.

wrobelda replied:
> It's LGPLV3. That might be a deal breaker for @benoitg because including that source here [makes libofx GPLV3](https://www.gnu.org/licenses/gpl-faq.html#AllCompatibility) instead of the current GPLV2+.

Actually, looking at the source code, the license is _"either version 2 of the License, or (at your option) any later version. ",_ which you already noticed by referring to it as GPLV2+. That makes libofx GPL3-compatible, which, in turn, LGPLV3 is compatible with, so I don't think we have a problem here?

EDIT: OK, I think I understand what you meant, i.e. that including the GPLV3-compatible code would automatically make that GPLV2+ license get upgraded to GPLV3, and as a result no longer leave end user with a choice of which version of GPL to use.

wrobelda replied:
> Indeed. The alternative is MIT-licensed https://github.com/takamin/win-c

I updated the code to use the aforementioned alternative.

jralls commented on CMakeLists.txt:
```diff
@@ -55,41 +55,13 @@ endif(MSVC)
 target_link_libraries(ofx ${OpenSP_LIBRARIES} ${ICONV_LIBRARIES})
 
 # ofxconnect
-find_package(CURL)
```
Best to leave all of the `find_package` calls in the root CMakeLists.txt, ideally grouped together, to make it easy for builders to determine all of the package's dependencies.

wrobelda replied:
I was debating between doing that when cleaning a bit up and indeed realized I should bring all the `find_package` calls back to main file, as there aren't too many.

jralls commented on cmake/modules/FindOpenSP.cmake:
```diff
@@ -0,0 +1,90 @@
+endif ()
+
+include(FindPackageHandleStandardArgs)
+find_package_handle_standard_args(OpenSP
+        FOUND_VAR
+        OpenSP_FOUND
+        REQUIRED_VARS
+        OpenSP_LIBRARY
+        OpenSP_INCLUDE_DIR
+        VERSION_VAR
+        OpenSP_VERSION_STRING
```
It's more ideomatic to put the designators and values on one line, i.e.
```
find_package_handle_standard_args(OpenSP
        FOUND_VAR OpenSP_FOUND
        REQUIRED_VARS OpenSP_LIBRARY OpenSP_INCLUDE_DIR
        VERSION_VAR OpenSP_VERSION_STRING
)
```

wrobelda committed:
Update copyrights.

FindOpenSP.cmake was written from scratch.

jralls commented on cmake/modules/FindOpenSP.cmake:
```diff
@@ -1,4 +1,3 @@
-# SPDX-FileCopyrightText: 2009 Cristian Oneț <onet.cristian@gmail.com>
 # SPDX-FileCopyrightText: 2022 Dawid Wróbel <me@dawidwrobel.com>
```
Still needs a license statement. If you're OK with GPL2-or-later just use 
```
/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
```
(Lifted from context.cpp).

wrobelda replied:
BSD 3 for compatibility with CMake (where this module was filed as a PR).

jralls replied:
You can license it as GPL here and as BSD 3-clause in Cmake if you want to. It's your work. Regardless you need to say here what license applies and if it's not the GPL V2 you need to add a copy of the other license in the top source directory.

wrobelda replied:
I was going to keep it the same until a decision is made on Kitware end whether or not to they would accept it to CMake. The idea was that in former case, any future changes made to CMake's copy would not be back-portable to libofx's. But  in that scenario I suppose we could just replace the whole copy altogether with vanilla, if needed — or remove it altogether and bump the `cmake_minimum_required`.

wrobelda committed:
FindOpenSP: add license header

wrobelda committed:
More fixes/improvements:
- install lib/bin in their respective CMakeLists
- rename Getopt::Getopt target

wrobelda committed:
Move find_package() calls back to root CMakeLists

wrobelda committed:
Use target_include_directories() instead include_directories()

wrobelda committed:
Use MIT getopt implementation for Windows

Source from: https://github.com/takamin/win-c

wrobelda committed:
Cleanup windows leftovers

wrobelda committed:
Declare library source and header files explicitly

wrobelda committed:
Use target_compile_definitions() instead add_definitions()

Also fix version reporting from command line

Stephen-Leibowitz reviewed (commented):
I think the two if statements always evaluate to true and can be removed.

ofx2qif/CMakeLists.txt
(line 6) if (ENABLE_OFX2QIF)

ofxdump/CMakeLists.txt
(line 7) if (ENABLE_OFXDUMP)

wrobelda commented:
> I think the two if statements always evaluate to true and can be removed.
> 
> ofx2qif/CMakeLists.txt (line 6) if (ENABLE_OFX2QIF)
> 
> ofxdump/CMakeLists.txt (line 7) if (ENABLE_OFXDUMP)

These are made as such to be features configurable from the command line, e.g.
```
$ cmake -DENABLE_OFX2QIF=OFF -DENABLE_OFXDUMP=OFF .
-- The following REQUIRED packages have not been found:

-- The following OPTIONAL packages have not been found:

-- The following features have been enabled:
 * ofxconnect, Enables ofxconnect utility which allows to test OFX Direct Connect

-- The following features have been disabled:
 * ofxdump, Enables ofxconnect utility which prints, in human readable form, everything the library understands about a file
 * ofx2qif, Enables OFX file to QIF (Quicken Interchange Format) file converter

-- Configuring done
-- Generating done
-- Build files have been written to: /Users/cromo/Documents/Sourcecode/libofx
```

wrobelda committed:
Update FindOpenSP module's license

wrobelda committed:
Fix FindOpenSP returning parent include dir

wrobelda committed:
Fix libofx.pc.in tokenization

wrobelda committed:
Use CMAKE_BINARY_DIR consistently

wrobelda committed:
Fix CMAKE_BINARY_DIR <> CMAKE_CURRENT_BINARY_DIR usage

wrobelda committed:
Define MAKEFILE_DTD_PATH properly

wrobelda committed:
Add DISABLE_OPENSP_MULTIBYTE option

wrobelda commented:
This should now be ready for a review. I tested it on each platform and it seems to be working fine, outside of Windows where I am most likely facing an unrelated issue (as discussed in #62)

One more issue with Windows I see is that `install(TARGETS tool name RUNTIME DESTINATION bin)` does not copy the `opensp.dll` on Windows (the lib gets copied, though). Not sure if this has something to do with my FindOpenSP that may need some extra magic to detect how the build was build (it does `add_library(OpenSP::OpenSP UNKNOWN IMPORTED)`, and the UNKNOWN is how cmake manual advertises its use for find modules), or with the `install` syntax itself. 

Last remaining task is the HAVE_GCC_VISIBILITY_EXTS handling. Do we need to take care of it at all anymore?

wrobelda committed:
Add getopt license

wrobelda committed:
Be verbose when looking up getopt implementation

wrobelda committed:
CMake support was rewritten entirely, drop unrelated copyrights

cstim commented:
As for GCC_VISIBILITY: This feature is supported by any gcc >= 4 (see here https://gcc.gnu.org/wiki/Visibility and the last bullet list specifically mentions this), hence I think you can safely assume that if this is gcc, the feature is supported. No need for the extra test compile that is in configure.ac.

For the (non-windows) gcc build, this should be kept, hence.

cstim commented on lib/CMakeLists.txt:
```diff
@@ -0,0 +1,53 @@
+
+add_library(ofx ${ofx_SRCS} ${ofx_HEADERS})
+target_link_libraries(ofx OpenSP::OpenSP)
+target_include_directories(ofx PRIVATE ${ofx_HEADERS} ${COMMON_INCLUDE_DIRS})
```
Change this line to remove `${ofx_HEADERS}` (the directive should only include directories, not files)

wrobelda replied:
Fixed.

cstim commented on lib/CMakeLists.txt:
```diff
@@ -0,0 +1,53 @@
+        win32.hh
+        )
+
+add_library(ofx ${ofx_SRCS} ${ofx_HEADERS})
```
Similar to John, I'm puzzled by the missing SHARED (or STATIC) option here. This means when building everything by default, you just get a static library but no shared one - which is different to the autotools system that would by default build the shared one and not the static one.

I had to read about BUILD_SHARED_LIBS in the cmake docs because I didn't know about this (relatively new?) option, and the rest of your CMake files here doesn't mention this. Hence it also doesn't show up as an option in any cmake GUI such as `ccmake` or similar. Is there a possibility where you could highlight that this option could be set? And also set teh default to BUILD_SHARED_LIBRARY=on, to keep the same default behaviour as in autotools?

wrobelda replied:
Resorting to `BUILD_SHARED_LIBS` instead of marking it explicitly as SHARED/STATIC is preferred because it allows wrappers like `vcpkg` to make that decision depending on the use case. But as you suggest, adding a user-facing option to enable/disable it explicitly is indeed a good practice. Fixed.

cstim reviewed (changes requested):
One minor bugfix in Cmake. And also the question about BUILD_SHARED_LIBS, potentially documenting it, and potentially setting its default to ON to keep the same default as with autotools.

Apart from that, I think it's good and a great step forward. Thanks a lot!

wrobelda committed:
Fix target_include_directories() usage

wrobelda committed:
BUILD_SHARED_LIBS by default

wrobelda commented:
@cstim note that this has 2-3 issues left to resolve, see the first comment for TODO list. That Windows issue is particularly important, but I am close to fixing it.

wrobelda committed:
Properly detect a non-const inbuffer in iconv()

Previously a _WIN32 presence was assumed as an indicator for iconv()
signature, which is wrong.

Fixes #51

wrobelda committed:
Export all symbols on Windows

wrobelda committed:
Add SP_MULTI_BYTE detection logic

wrobelda commented:
> For the (non-windows) gcc build, this should be kept, hence.

@cstim the problem with it is that it is also guarded by `IN_LIBOFX`, which seems to be always enabled? If you don't mind, I'll leave it to you to deal with, especially that it is non-critical to get the lib and tools compiling.

cstim commented on config.h.cmake:
```diff
@@ -0,0 +1,17 @@
+/* Version number of package */
+#define VERSION "@LIBOFX_VERSION_RELEASE_STRING@"
+
+#cmakedefine _WIN32 1
```
I still don't understand why _WIN32 is being defined in our build system. If we're on windows, it exists anyway, see https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170 for MSVC and https://blog.kowalczyk.info/article/j/guide-to-predefined-macros-in-c-compilers-gcc-clang-msvc-etc..html for an overview for all systems.

wrobelda replied:
Ah yes, forgot to remove it.

wrobelda replied:
Fixed.

cstim committed:
For gcc compiler, enable the visibility=hidden flag using cmake's CXX_VISIBILITY_PRESET

cstim commented:
With the VISIBILITY_HIDDEN, I pushed a commit that (I think) solves this as well, see https://github.com/libofx/libofx/commit/8908928348b041977c014b18d85283d905a40026

To verify yourself: Run `nm -C  build-cmake/lib/libofx.so   | grep ' T ' | wc ` before and after. Before, it mentions 159 symbols to be exported (marked with " T "), after, it mentions 16 symbols to be exported (the others are marked with " t ", lower-case t now).

cstim commented:
And for IN_LIBOFX: This is a define to be used inside the library's header files to decide whether we are currently compiling the library itself vs. we are merely including the header to use the library in another (downstream) executable, such as in ofxdump. The pattern occurs quite often, and is helpful in the headers that are included from the other executables but are also used while building the library itself. I've defined it in my commit as well.

wrobelda committed:
Remove leftover _WIN32 #cmakedefine

wrobelda committed:
Add CMakeCache.txt to .gitignore

wrobelda committed:
Reuse glibc implementation of getopt

- Remove Windows-specific code
- Use glibc copy used previously
- Make wit windows-compatible

wrobelda committed:
Fix ofxconnect target

wrobelda commented:
@cstim can you please check if  c280f94d4c4fc50de2539e8f352bc0fe49f14641 looks OK on your end? Specifically that `configure.ac` change in it.

wrobelda committed:
Add vcpkg manifest file

wrobelda committed:
Fix finding external GetOpt library

wrobelda committed:
Improve vcpkg integration

Enabling any of the tools makes vcpkg automatically download all their required dependencies.

wrobelda committed:
Reformat CMake code

cstim commented on configure.ac:
```diff
@@ -274,9 +274,9 @@ AC_DEFUN([adl_FUNC_GETOPT_LONG],
+      AC_CONFIG_LIBOBJ_DIR(getopt)
       AC_LIBOBJ(getopt)
       AC_LIBOBJ(getopt1)
-      AC_CONFIG_LINKS([lib/getopt.h:lib/gnugetopt.h])])])])
```
You're accidentally deleting a few too much closing parentheses here. It still needs to have `])])])` (three square and three round parentheses).

wrobelda replied:
Fixed both.

cstim reviewed (changes requested):
The fixup in configure.ac is needed.

And I would keep the default to build the tools, because this will also act as a test-cases that linking with the library works fine.

wrobelda committed:
Update READNE

- add notes on CMake
- restructe the contents, add more headers
- move FAQ contents to it

wrobelda commented:
@cstim I am still running into issues with `SP_MULTI_BYTE` on Windows, and this is caused by a wrong `config.h` getting processed. It's hard to solve this without moving it to some subfolder (`inc\`, presumably) and changing all the includes to load it?

EDIT: hopefully fixed with https://github.com/libofx/libofx/pull/63/commits/98a8494b3fb135d170e05e718e02a76d4190320d

wrobelda committed:
Actually fix SP_MULTI_BYTE on Windows

cstim commented:
The autotools build system gets another issue now:
`Makefile.am: error: required file './README' not found`
Changing the filename doc_DOCS variable is NOT sufficient. I'd say just leave the README at the old filename and keep the cleanups for a later release.

Anyway, I will leave for vacation now. You would need to get final approval by someone else or wait until Aug 07 when I can run the build myself again. Have fun!

wrobelda commented:
> The autotools build system gets another issue now: `Makefile.am: error: required file './README' not found` Changing the filename doc_DOCS variable is NOT sufficient. I'd say just leave the README at the old filename and keep the cleanups for a later release.

Adding the `.md` helps IDEs (e.g. CLion) detect that it is a markdown file, so definitely something worth doing, but absolutely can be left for later. I can actually have a separate PR Draft readied for that.

> Anyway, I will leave for vacation now. You would need to get final approval by someone else or wait until Aug 07 when I can run the build myself again. Have fun!

Cool, not a problem. This is basically ready now, from my perspective. The only remaining issue is getting it build dynamically on Windows, but that wasn't supported before, either. Enjoy your vacation!

EDIT: building a shared lib on Windows works now, too.

wrobelda committed:
Move ofxconnect's READMEs to main README.md

wrobelda committed:
Rename README.md back to README to satisfy autotools

This will be renamed again at a later time

wrobelda committed:
REAME: some formatting, style

wrobelda committed:
Add Support for bulding shared lib on Windows

wrobelda committed:
Disable CRT Secure warnings on Windows

wrobelda commented:
@jralls if you don't mind checking the b8b3e9f96ce57897b1c0d7f64462259f45d0218c, I would appreciate it

wrobelda committed:
README: add notes on OpenSP_INCLUDE_DIR and OpenSP_LIBRARY

wrobelda committed:
README: phrasing, paragraphs, etc.

wrobelda committed:
README: some formatting

wrobelda committed:
README: troubleshooting ofxconnect failing to build with vcpkg due to libxml++ ABI mismatch

wrobelda committed:
CMakeLists.txt: formatting

jralls commented on CMakeLists.txt:
```diff
@@ -0,0 +1,201 @@
+
+    find_package(PkgConfig)
+    if (PkgConfig_FOUND)
+        pkg_check_modules(LIBXMLPP IMPORTED_TARGET libxml++-2.6>=2.6)
```
needs 
```
if (NOT LIBXMLPP_FOUND)
   set(ENABLE_OFXCONNECT OFF)
endif()
```
Otherwise if it's missing generation fails with
```
CMake Error at ofxconnect/CMakeLists.txt:14 (add_executable):
  Target "ofxconnect" links to target "PkgConfig::LIBXMLPP" but the target
  was not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?
```

wrobelda replied:
> One thing that's missing is a dist target: autotools provides one by default, cmake doesn't. The main purpose is to have a working configure script that doesn't need autotools installed. 

Why would you insist on generating that tarball with CMake instead of autotools, then? Especially that it was agreed that support for autotools (and a pre-generated configure script) would be dropped around +1 release, so all this effort would be futile.

> The other reason is to have a tarball that's named the way the FOSS world expects instead of Github's version-only vX.Y.tar.gz with no project name.

I second my opinion that it seems unnecessary. Outside of your quoted example of Gnucash, I have not once seen a CMake-only software that did that just for that purpose. 

But most importantly, pretty much all of the software distribution systems I am aware of can work with Git directly, either via a tag or branch, accompanied by a SHA, and don't need a hosted tarball anymore. In any other case, I'd argue that a tarball generated by Github would suffice. Adding extra scripting sounds very un-K.I.S.S. to me.

wrobelda replied:
> needs
>
> if (NOT LIBXMLPP_FOUND)
>    set(ENABLE_OFXCONNECT OFF)
> endif()

I strongly disagree on that. If `ENABLE_OFXCONNECT` is enabled and a required dependency is not found, then it should fail as it does instead of transparently changing a user-facing configurable feature. Such implicit behavior is easy to miss, especially when used in automated builds. I can easily imagine `libxml` somehow getting removed from the underlying CI/CD system, resulting in a successful build with a missing ofxconnect binary. KMyMoney, for example, does include them for users to help with investigation when needed.

jralls replied:
> But most importantly, pretty much all of the software distribution systems I am aware of can work with Git directly, either via a tag or branch, accompanied by a SHA, and don't need a hosted tarball anymore. In any other case, I'd argue that a tarball generated by Github would suffice. Adding extra scripting sounds very un-K.I.S.S. to me.

OK, but it's a change in behavior and means that @benoitg and @cstim have to change their release process. Whether they want to do that is up to them.

>  If ENABLE_OFXCONNECT is enabled and a required dependency is not found, then it should fail as it does instead of transparently changing a user-facing configurable feature. 

That's a change in behavior, configure.ac does:
```
PKG_CHECK_MODULES(LIBCURL, libcurl >= 7.9.7,
        [libcurl_available=yes],
        [libcurl_available=no
         AC_MSG_WARN([libcurl is not available. ofxconnect (Direct connect samples) will NOT be built.])
        ])

PKG_CHECK_MODULES(LIBXMLPP,libxml++-2.6 >= 2.6,
	[
	AC_DEFINE(HAVE_LIBXMLPP, 1, [Defined if libxml++ is available])
	have_libxmlpp=yes],
        [AC_MSG_WARN([libxml++ 2.6 is not available. ofxconnect (Direct connect samples) will NOT be built.])
	have_libxmlpp=no])
```

A reasonable alternative would be to default `ENABLE_OFXCONNECT` to `OFF` so that a default build doesn't fail for an optional component. You could also provide a more verbose explanatory error message suggesting that the failure is because of an optional component that's enabled by default and that the builder can disable it instead of installing the dependencies.

A more serious concern and a bit off topic: Does ofxconnect still work? ofxpartner.cpp has a hard-coded url `http://moneycentral.msn.com/money/2005`. `moneycentral.msn.com` isn't in the DNS anymore. That aside, GnuCash uses [AQBanking](https://www.aquamaniac.de) for OFX-DirectConnect support and I've worked with its developer several times in the last few years to deal with changes Intuit has made to the way the banks' servers authenticate. The [last functional commit](https://github.com/libofx/libofx/commit/74e8e94e1abb2bb316239e410b33822bb4d91fa1) to ofxconnect was in 2007,  more than 15 years ago. Everything since has been just to keep it building.

wrobelda replied:
> OK, but it's a change in behavior and means that @benoitg and @cstim have to change their release process. 

Since the tarballs get published as is on Github already — albeit without the autootool scripting — hopefully it won't be much of a change once autotool scripting gets dropped. 

> That's a change in behavior, configure.ac does

Building of the tools wasn't even configurable until very recently (see: https://github.com/libofx/libofx/commit/f4f7c473beae782e15bbd3fdad1fd4543dd8c3ff), so the original behavior of `./configure` script that you refer to seems reasonable. All the tools are configurable separately in CMake, though, so it isn't much of an issue as they can be disabled at will when dependencies are missing.

> A reasonable alternative would be to default ENABLE_OFXCONNECT to OFF

That's what I initially proposed, but @cstim suggested otherwise: https://github.com/libofx/libofx/pull/63#pullrequestreview-1048681174

> You could also provide a more verbose explanatory error message suggesting that the failure is because of an optional component that's enabled by default and that the builder can disable it instead of installing the dependencies.

Actually, I thought that `feature_summary(WHAT REQUIRED_PACKAGES_NOT_FOUND` would make it obvious enough as to what caused what, without custom messaging. But it turns out that it's not even shown, since when first missing but otherwise required package is found, CMake fails on the spot. Very puzzling, trying to figure it out.

And there's that, too: https://gitlab.kitware.com/cmake/cmake/-/issues/22689

> A more serious concern and a bit off topic: Does ofxconnect still work? 

Actually I was wondering about the same — see the `Is ofxconnect's test-privateserver.sh still functional?` in my first comment that touches on that.

jralls replied:
>Is ofxconnect's test-privateserver.sh still functional?

It [requires a gpg key](https://github.com/libofx/libofx/blob/master/ofxconnect/README.privateserver) to test it or even to see what it does. I'll try the email address in that readme, but I don't have much hope. But it's also orthogonal to the real problem because even if that server is still active it isn't running what real banks do. If that private server is still up it probably works with ofxconnect, but that provides no indication of whether ofxconnect will connect to a real bank.

wrobelda replied:
> Actually, I thought that feature_summary(WHAT REQUIRED_PACKAGES_NOT_FOUND would make it obvious enough as to what caused what, without custom messaging. But it turns out that it's not even shown, since when first missing but otherwise required package is found, CMake fails on the spot. Very puzzling, trying to figure it out.

OK, I sorted that all out with 7a54725f8276d49d06a6a08037d21b25a0c69113, which shows a meaningful error when a package is missing, alongside the feature summary. It doesn't work for `pkg_check_modules`, though.

wrobelda replied:
> It [requires a gpg key](https://github.com/libofx/libofx/blob/master/ofxconnect/README.privateserver) to test it or even to see what it does. I'll try the email address in that readme, but I don't have much hope. But it's also orthogonal to the real problem because even if that server is still active it isn't running what real banks do. If that private server is still up it probably works with ofxconnect, but that provides no indication of whether ofxconnect will connect to a real bank.

I suppose only  @benoitg and @cstim  can answer that. That having said, KMyMoney uses LibOFX alongside AqBanking and both can be used for OFX Direct Connect, which I personally do for my banks randomly, so I'd wager that Direct Connect is still functional in LibOFX — at least for Amex in my case.

benoitg replied:
> > OK, but it's a change in behavior and means that @benoitg and @cstim have to change their release process.
> 
> Since the tarballs get published as is on Github already — albeit without the autootool scripting — hopefully it won't be much of a change once autotool scripting gets dropped.

I really don't mind.  Your work to modernize the build infrastructure to improve cross-platform building is a large piece of work.  Tweaking the (currently manual) release process is comparatively trivial.

jralls commented on CMakeLists.txt:
```diff
@@ -0,0 +1,201 @@
+endif ()
+
+option(ENABLE_OFXCONNECT "Enable ofxconnect tool" ON)
+add_feature_info("ofxconnect" ENABLE_OFXCONNECT "Enables ofxconnect utility which allows to test OFX Direct Connect")
+if (ENABLE_OFXCONNECT)
+    list(APPEND VCPKG_MANIFEST_FEATURES "ofxconnect")
+endif ()
```
Move this and the corresponding ICONV lines to the end of the respective dependency check sections so that the manifest and feature summary correctly reflect them being disabled if the dependencies aren't satisfied.

wrobelda replied:
> Move this and the corresponding ICONV lines to the end

This was purposefully moved to the front of `project(libofx)`, as required by vcpkg:
https://vcpkg.io/en/docs/users/buildsystems/cmake-integration.html#settings-reference
https://vcpkg.io/en/docs/users/buildsystems/cmake-integration.html#vcpkg_manifest_features

> correctly reflect them being disabled if the dependencies aren't satisfied.

None of the dependencies are optional, so as discussed, any that are missing will fail the build anyway — that is unless a feature requiring them was explicitly disabled, in which case the feature summary will reflect the situation correctly.

jralls replied:
Can you test that from the command line? I just did with my USAA account and got "Request unsuccessful".  I see when I dump the request to stdout instead of sending it that ofxconnect sends VERSION:102, <APPID>QWIN and <APPVER>1400 and that it has no provision for a <CLIENTUID>. USAA requires VERSION:103, <APPID>QMOFX and <APPVER>2300 along with a <CLIENTUID> which it provides as part of getting credentials. If any of that's configurable in ofxconnect it doesn't show up in `ofxconnect --help`.  Maybe AMEX is more liberal?

jralls replied:
The `--bank-list`, `--bank-fipid`, `--bank-services`, and `--allsupport` functions return parse errors because http://moneycentral.msn.com/money/2005/mnynet/service/ols/filist.aspx?SKU=3&VER=6 redirects to https://www.msn.com/money/. ofxserver.cpp barfs because it doesn't tell curl to follow the redirect, but even if it did the page is just MSN's financial landing page.

It could be rewritten to use https://ofxhome.com, but that's [on borrowed time](http://www.ofxhome.com/ofxforum/viewtopic.php?id=49843). I don't think there's a suitable replacement.

Stephen-Leibowitz replied:
There has been a recent announcement, “[New Ownership of OFXHome](https://www.ofxhome.com/ofxforum/viewtopic.php?pid=109005).”

jralls commented on ofxconnect/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+        )
+
+# required by libxml++
+set(CMAKE_CXX_STANDARD 11)
```
Make this
```
if (CMAKE_CXX_STANDARD LESS 11)
    set(CMAKE_CXX_STANDARD 11)
endif()
```
So that one can set a newer standard if necessary. For example to compile with a recent GLib one must build glibmm with `-Dbuild_deprecated_api=true`, copy glibmm-2.68.pc to glibmm-2.4.pc, and build libxml++-2.42 and ofxconnect with -std=c++-17 because glibmm uses `is_foo_v` type traits. (One must also replace all instances in libxml++ of `std::auto_ptr` with `std::unique_ptr` and change a couple of `std::string` to `Glib::ustring` in ofxconnect, but that's way outside the scope of this PR).

wrobelda replied:
Good point, thanks. Done.

jralls reviewed (changes requested):
The commit you requested went away when you force-pushed a couple of hours later, so I reviewed the whole PR.

One thing that's missing is a `dist` target: autotools provides one by default, cmake doesn't. The main purpose is to have a working configure script that doesn't need autotools installed. That's not an option for cmake. The other reason is to have a tarball that's named the way the FOSS world expects instead of Github's version-only vX.Y.tar.gz with no project name.  It's up to @cstim whether he wants one. If he does, feel free to adapt https://github.com/Gnucash/gnucash/blob/maint/common/cmake_modules/MakeDist.cmake.

wrobelda committed:
Don't force the CXX standard to 11 if it was set higher than that

wrobelda committed:
PKGCONFIG and LIBXMPP are required for OFXCONNECT

wrobelda committed:
Only include win32 code on Windows

wrobelda committed:
Use set_package_properties to set REQUIRED on packages required

cstim reviewed (approved):
I think we're good to go.

The config.h changes are not dropping anything that is still relevant, so it's fine.

The "make dist" target is no more important, too - the cmake target would be fine just as well or any other process - we'll figure soon.

wrobelda commented:
> I think we're good to go.

Feel free to squash everything. Thanks!
